pkg/kubelet/cm/dra/plugin/noderesources.go

```diff
@@ -0,0 +1,449 @@
+// with the resource information provided by plugins. Without it,
+// the controller is inactive.
+func startNodeResourcesController(ctx context.Context, kubeClient kubernetes.Interface, nodeName string) *nodeResourcesController {
+	if kubeClient == nil {
+		return nil
+	}
+
+	logger := klog.FromContext(ctx)
+	logger = klog.LoggerWithName(logger, "node resources controller")
+	ctx = klog.NewContext(ctx, logger)
+
+	c := &nodeResourcesController{
+		ctx:           ctx,
+		kubeClient:    kubeClient,
+		nodeName:      nodeName,
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "node_resource_slices"),
+		activePlugins: make(map[string]*activePlugin),
+	}
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		c.run(ctx)
+	}()
+
+	return c
+}
+
+// waitForStop blocks until all background activity spawned by
+// the controller has stopped. The context passed to start must
+// be canceled for that to happen.
+//
+// Not needed at the moment, but if it was, this is what it would
+// look like...
+// func (c *nodeResourcesController) waitForStop() {
+// 	if c == nil {
+// 		return
+// 	}
+//
+// 	c.wg.Wait()
+// }
+
+// addPlugin is called whenever a plugin has been (re-)registered.
+func (c *nodeResourcesController) addPlugin(driverName string, pluginInstance *plugin) {
+	if c == nil {
+		return
+	}
+
+	klog.FromContext(c.ctx).V(2).Info("Adding plugin", "driverName", driverName)
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if active := c.activePlugins[driverName]; active != nil {
+		active.cancel(errors.New("plugin has re-registered"))
+	}
+	active := &activePlugin{}
+	cancelCtx, cancel := context.WithCancelCause(c.ctx)
+	active.cancel = cancel
+	c.activePlugins[driverName] = active
+	c.queue.Add(driverName)
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		c.monitorPlugin(cancelCtx, active, driverName, pluginInstance)
+	}()
+}
+
+// removePlugin is called whenever a plugin has been unregistered.
+func (c *nodeResourcesController) removePlugin(driverName string) {
+	if c == nil {
+		return
+	}
+
+	klog.FromContext(c.ctx).V(2).Info("Removing plugin", "driverName", driverName)
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if active, ok := c.activePlugins[driverName]; ok {
+		active.cancel(errors.New("plugin has unregistered"))
+		delete(c.activePlugins, driverName)
+		c.queue.Add(driverName)
+	}
+}
+
+func (c *nodeResourcesController) monitorPlugin(ctx context.Context, active *activePlugin, driverName string, pluginInstance *plugin) {
+	logger := klog.FromContext(ctx)
+	logger = klog.LoggerWithValues(logger, "driverName", driverName)
+	logger.Info("Starting to monitor node resources of the plugin")
+	defer func() {
+		r := recover()
+		logger.Info("Stopping to monitor node resources of the plugin", "reason", context.Cause(ctx), "err", ctx.Err(), "recover", r)
+	}()
+
+	// Keep trying until canceled.
+	for ctx.Err() == nil {
+		logger.V(5).Info("Calling NodeResources")
```

Comment:
Update string to new api call name

Reply:
Missed it in this pass, will be in next push.